src/Facility.LanguageServer/FsdDefinitionUtility.cs

```diff
@@ -17,6 +17,53 @@ public static ServiceMemberInfo GetMemberReferencedAtPosition(this ServiceInfo s
 				select service.FindMember(name)).FirstOrDefault();
 		}
 
+		public static IEnumerable<ServicePart> GetReferencedServicePartsAtPosition(this ServiceInfo service, Position requestPosition)
+		{
+			var members = service.GetDescendants().OfType<ServiceMemberInfo>().ToList().AsReadOnly();
+
+			// memberNameAtCursor will be null if the cursor is not on a member name.
+			var memberNameAtCursor = members
+				.Select(member =>
+				{
+					var part = member.GetPart(ServicePartKind.Name);
+					var name = member.Name;
+
+					return (part, name);
+				})
+				.Where(x => x.part != null && requestPosition >= x.part.Position && requestPosition < x.part.EndPosition)
+				.Select(x => x.name)
+				.FirstOrDefault();
+
+			var fields = service.GetDescendants().OfType<ServiceFieldInfo>().ToList().AsReadOnly();
+
+			// fieldTypeNameAtCursor will be null if the cursor is not on a field type name.
+			var fieldTypeNameAtCursor = fields
+				.Select(field =>
+				{
+					var part = field.GetPart(ServicePartKind.TypeName);
+					var typeName = service.GetFieldTypeName(field);
+
+					return (part, typeName);
+				})
+				.Where(x => x.part != null && requestPosition >= x.part.Position && requestPosition < x.part.EndPosition)
+				.Select(x => x.typeName)
+				.FirstOrDefault();
+
+			return fields
+				.Select(field =>
+				{
+					var part = field.GetPart(ServicePartKind.TypeName);
+					var typeName = service.GetFieldTypeName(field);
+
+					var type = service.GetFieldType(field);
+					var memberTypeName = type?.GetMemberTypeName();
+
+					return (part, memberTypeName, typeName);
+				})
+				.Where(x => x.part != null && ((memberNameAtCursor != null && x.memberTypeName == memberNameAtCursor) || x.typeName == fieldTypeNameAtCursor))
+				.Select(x => x.part);
+		}
```

Comment:
This could be rewritten using the query syntax if that is preferred. I originally used the method syntax because it was easier to debug.

```suggestion
		public static IEnumerable<ServicePart> GetReferencedServicePartsAtPosition(this ServiceInfo service, Position requestPosition)
		{
			// memberNameAtCursor will be null if the cursor is not on a member name.
			var memberNameAtCursor = (
				from member in service.GetDescendants().OfType<ServiceMemberInfo>()
				let part = member.GetPart(ServicePartKind.Name)
				let memberName = member.Name
				where part != null && requestPosition >= part.Position && requestPosition < part.EndPosition
				select memberName).FirstOrDefault();

			var fields = service.GetDescendants().OfType<ServiceFieldInfo>().ToList().AsReadOnly();

			// fieldTypeNameAtCursor will be null if the cursor is not on a member name.
			var fieldTypeNameAtCursor = (
				from field in fields
				let part = field.GetPart(ServicePartKind.TypeName)
				let typeName = service.GetFieldTypeName(field)
				where part != null && requestPosition >= part.Position && requestPosition < part.EndPosition
				select typeName).FirstOrDefault();

			return
				from field in fields
				let part = field.GetPart(ServicePartKind.TypeName)
				let typeName = service.GetFieldTypeName(field)
				let type = service.GetFieldType(field)
				let memberTypeName = type?.GetMemberTypeName()
				where part != null && ((memberNameAtCursor != null && memberTypeName == memberNameAtCursor) || typeName == fieldTypeNameAtCursor)
				select part;
		}
```

Reply:
That does read better, especially if you move the `where part != null` to immediately after `let part =` lines, but I don't feel strongly about it.